pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -886,6 +888,35 @@ public void redeliverUnacknowledgedMessages() {
     }
 
     @Override
+    public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
+        if (!conf.isPerMessageRedeliverUnacknowledged()
+                || messageIds.size() > conf.getMaxRedeliverUnacknowledgedMessagesBatch()) {
+            redeliverUnacknowledgedMessages();
+            return;
+        }
+        ClientCnx cnx = cnx();
+        if (isConnected() && cnx.getRemoteEndpointProtocolVersion() >= ProtocolVersion.v2.getNumber()) {
+            MessageIdData.Builder builder = MessageIdData.newBuilder();
+            List<MessageIdData> messageIdDatas = messageIds.stream()
+                    .map(messageId -> {
+                        builder.setPartition(messageId.getPartitionIndex());
```

Comment:
I did this to avoid generating garbage or cpu time, since we're replacing the same fields on every iteration there should be no problems with reusing the same builder instance and should save calls to recycler.